### ISSUE TITLE
Mixing Numpy's arrays and PyTorch tensors

### ISSUE THREAD

**Issue by Atcold** [pytorch/pytorch#46829]
## 🐛 Bug

`t + a != a + t`, where `t` is a tensor and `a` is an array.

## To Reproduce

```python
> t = torch.arange(3.)
> a = -arange(3.)
> print(t + a)
> print(a + t)

tensor([0., 0., 0.], dtype=torch.float64)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1279-6264a51fb86e> in <module>
      2 a = -arange(3.)
      3 print(t + a)
----> 4 print(a + t)

TypeError: unsupported operand type(s) for +: 'numpy.ndarray' and 'Tensor'
```

## Expected behavior

`t + a == a + t`.

## Environment

 - PyTorch Version (e.g., 1.0): 1.5.1

**Comment by rgommers** [comment]
This is a natural consequence of NumPy being more forgiving than PyTorch on what kinds of things it accepts and tries to convert to `ndarray`s. 

Note that `t + a` will also error if `t` doesn't live on CPU.

I'm not sure there's a behaviour change to be made here. If there is, it'd be returning `NotImplemented` from `Tensor.__add__` rather than raising a `TypeError`. But I suspect that that then has to be done for many other dunder methods, and may have some other impacts that need to be assessed.

**Comment by Atcold** [comment]
I think we can assume that if we're mixing tensors and arrays, we'd want the output to be arrays.
Like, when plotting stuff, without having to `.numpy()` is quite convenient. Still, it breaks sometimes.

**Comment by rgommers** [comment]
> I think we can assume that if we're mixing tensors and arrays, we'd want the output to be arrays.

That really depends on what one is doing - and there's also a good chance it's user error. 

If you're just talking about operators rather than function calls: `+`, `__add__`, `torch.add`/`numpy.add` are typically all the same implementation. `torch.add(t, a)` should give a tensor, while you're probably arguing that `t + a` should give an array.

In the face of ambiguity, refuse the temptation to guess I'd say.

> Like, when plotting stuff, without having to `.numpy()` is quite convenient.

This should be solved on the plotting library side. If, e.g., Matplotlib only uses NumPy arrays internally, then it should convert other types of arrays/tensors to numpy arrays. There's discussions on how to make this easier to do in general (e.g. `__array__(force=True)`, without Matplotlib et al. having to know that PyTorch has special syntax to detach/move/convert-to-numpy.

**Comment by rgommers** [comment]
I think this is a wontfix. @mruberry?

**Comment by mruberry** [comment]
There's probably not a great fix for this. I marked it as "needs design."

**Comment by rgommers** [comment]
Oh wait, there is something to do here, because `__add__` / `__radd__` behave differently from the `add` function (which is symmetric in both pytorch and numpy):
```
>>> np.add(a, t)                                                                                                                           
tensor([0., 0., 0.], dtype=torch.float64)
>>> np.add(t, a)                                                                                                                           
tensor([0., 0., 0.], dtype=torch.float64)

>>> torch.add(a, t)                                                                                                                        
Traceback (most recent call last):
  File "<ipython-input-17-4227e219aca5>", line 1, in <module>
    torch.add(a, t)
TypeError: add(): argument 'input' (position 1) must be Tensor, not numpy.ndarray

>>> torch.add(t, a)                                                                                                                        
Traceback (most recent call last):
  File "<ipython-input-18-3351e32de46b>", line 1, in <module>
    torch.add(t, a)
TypeError: add(): argument 'other' (position 2) must be Tensor, not numpy.ndarray
```

Note that `np.add` returns a `Tensor` by design, because `Tensor.__array_wrap__` is defined in `torch/tensor.py` to achieve exactly that.

**Comment by rgommers** [comment]
NumPy won't change its behaviour, so I think the only way to do something about this is for PyTorch to start accepting NumPy arrays in all its functions, which I don't think we'd want to do.

**Comment by rsokl** [comment]
Could you give tensors a higher [`__array_priority__`](https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_priority__)?

**Comment by rgommers** [comment]
> Could you give tensors a higher [`__array_priority__`](https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_priority__)?

Higher than what other array/tensor type? `torch.Tensor` uses 1000, `np.ndarray` uses 0 right now.

**Comment by rsokl** [comment]
Sorry, I got mixed up when reading this and had been thinking that it was the tensors that were coercing the arrays to cooperate (and they _are_, but just not until numpy tries to call a ufunc on the tensor).

Please disregard my comment.
